hedera-node/hedera-app/src/main/java/com/hedera/node/app/uploader/MinioBucketUploader.java

```diff
@@ -0,0 +1,206 @@
+        this.bucketName = completeBucketConfigs.getFirst().bucketName();
+        this.provider = BucketProvider.valueOf(completeBucketConfigs.getFirst().provider());
+        this.maxRetryAttempts = configProvider
+                .getConfiguration()
+                .getConfigData(BlockStreamConfig.class)
+                .uploadRetryAttempts();
+        this.minioClients = MinioClientFactory.createClients(bucketConfigurationManager.getCompleteBucketConfigs());
+    }
+
+    private String calculateMD5Hash(Path filePath) throws IOException, NoSuchAlgorithmException {
+        // Calculate MD5 hash
+        byte[] fileBytes = Files.readAllBytes(filePath); // Read the file content as bytes
+        MessageDigest md = MessageDigest.getInstance("MD5");
+        byte[] md5Bytes = md.digest(fileBytes);
+        return Base64.getEncoder().encodeToString(md5Bytes);
+    }
+
+    @Override
+    public CompletableFuture<Void> uploadBlock(Path blockPath) {
+        return CompletableFuture.runAsync(
+                () -> {
+                    if (!Files.exists(blockPath)) {
+                        throw new IllegalArgumentException("Block path does not exist: " + blockPath);
+                    }
+                    String fileName = blockPath.getFileName().toString();
+                    String objectKey = fileName.endsWith(".blk.gz")
+                            ? fileName.replaceAll("[^\\d]", "") // Extract numeric part
+                            : "";
+                    try {
+                        // First check if object already exists
+                        if (blockExistsOnCloud(objectKey)) {
+                            String existingMd5 = getBlockMd5Internal(objectKey);
+                            if (existingMd5.equals(calculateMD5Hash(blockPath))) {
+                                logger.debug("Block {} already exists with matching MD5", objectKey);
+                                return;
+                            }
+                            throw new HashMismatchException(objectKey, provider.toString());
+                        }
+                        // Upload with retry logic
+                        RetryUtils.withRetry(
+                                () -> {
+                                    minioClients
+                                            .getFirst()
+                                            .uploadObject(UploadObjectArgs.builder()
+                                                    .bucket(bucketName)
+                                                    .object(objectKey)
+                                                    .filename(blockPath.toString())
+                                                    .contentType("application/octet-stream")
+                                                    .build());
+                                    return null;
+                                },
+                                maxRetryAttempts > 0 ? maxRetryAttempts : 3);
```

Comment:
We already have the default set to 3 in `BlockStreamConfig`, do we need this conditional here?